modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/stored/TransportPutStoredSearchTemplateAction.java

```diff
@@ -0,0 +1,116 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.script.mustache.stored;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.master.TransportMasterNodeAction;
+import org.elasticsearch.cluster.AckedClusterStateUpdateTask;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.block.ClusterBlockException;
+import org.elasticsearch.cluster.block.ClusterBlockLevel;
+import org.elasticsearch.cluster.metadata.IndexNameExpressionResolver;
+import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.script.ScriptException;
+import org.elasticsearch.script.ScriptMetaData;
+import org.elasticsearch.script.ScriptService;
+import org.elasticsearch.script.StoredScriptSource;
+import org.elasticsearch.script.TemplateService;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+public class TransportPutStoredSearchTemplateAction extends TransportMasterNodeAction<
+        PutStoredTemplateRequest, PutStoredTemplateResponse> {
+
+    private final TemplateService templateService;
+    private final int maxScriptSizeInBytes;
+
+    @Inject
+    public TransportPutStoredSearchTemplateAction(Settings settings,
+            TransportService transportService, ClusterService clusterService, ThreadPool threadPool,
+            ActionFilters actionFilters, IndexNameExpressionResolver indexNameExpressionResolver,
+            TemplateService templateService) {
+        super(settings, PutStoredTemplateAction.NAME, transportService, clusterService,
+                threadPool, actionFilters, indexNameExpressionResolver,
+                PutStoredTemplateRequest::new);
+        this.templateService = templateService;
+        maxScriptSizeInBytes = ScriptService.SCRIPT_MAX_SIZE_IN_BYTES.get(settings);
+    }
+
+    @Override
+    protected String executor() {
+        return ThreadPool.Names.SAME;
+    }
+
+    @Override
+    protected PutStoredTemplateResponse newResponse() {
+        return new PutStoredTemplateResponse();
+    }
+
+    @Override
+    protected void masterOperation(PutStoredTemplateRequest request, ClusterState state,
+            ActionListener<PutStoredTemplateResponse> listener) throws Exception {
+        if (request.content().length() > maxScriptSizeInBytes) {
```

Comment:
After this change ScriptService doesn't get involved in stored templates.

ScriptService does this check for scripts, yes, but I couldn't mirror that behavior for templates because these requests are in lang-mustache and TemplateService is in core.

I can do a few things:
1. Move the template requests to core. We'd talked about not wanting to do that but I'm not 100% clear on the reasoning.
2. Make a `StoredTemplatesService` that does all the stored template work and leave it in lang-mustache.